module/bdev/aio/bdev_aio.c

```diff
@@ -120,46 +120,135 @@ static struct spdk_bdev_module aio_if = {
-	fd = open(disk->filename, O_RDWR | O_DIRECT);
+	fd = open(disk->filename, (O_RDWR | O_DIRECT));
 	if (fd < 0) {
 		/* Try without O_DIRECT for non-disk files */
 		fd = open(disk->filename, O_RDWR);
 		if (fd < 0) {
-			SPDK_ERRLOG("open() failed (file:%s), errno %d: %s\n",
-				    disk->filename, errno, spdk_strerror(errno));
+			int rc = errno;
+
+			SPDK_ERRLOG("open(%s) failed, errno %d: %s\n",
+				    disk->filename, rc, spdk_strerror(rc));
 			disk->fd = -1;
-			return -1;
+			return ERROR(rc);
 		}
 	}
 
 	disk->fd = fd;
-
 	return 0;
 }
 
+/* zero region of a (sparse) file by creating a hole */
 static int
-bdev_aio_close(struct file_disk *disk)
+aio_range_fallocate_sparse(int fd, uint64_t range[])
+{
+	return fallocate(fd, (FALLOC_FL_PUNCH_HOLE | FALLOC_FL_KEEP_SIZE), range[0], range[1]);
+}
+
+/* zero region of a file by writing zeros */
+static int
+aio_range_fallocate_zero(int fd, uint64_t range[])
 {
+	return fallocate(fd, (FALLOC_FL_ZERO_RANGE | FALLOC_FL_KEEP_SIZE), range[0], range[1]);
+}
+
+/* unmap region of a device */
+static int
+aio_range_unmap_discard(int fd, uint64_t range[])
+{
+	return ioctl(fd, BLKDISCARD, range);
+}
+
+/* zero region of a device */
+static int
+aio_range_unmap_zero(int fd, uint64_t range[])
+{
+	return ioctl(fd, BLKZEROOUT, range);
+}
+
+static int
+set_aio_range_functions(struct file_disk *disk)
+{
+	struct stat st;
 	int rc;
 
-	if (disk->fd == -1) {
+	if (fstat(disk->fd, &st) < 0) {
+		rc = errno;
+		SPDK_ERRLOG("fstat(%s) failed, errno %d: %s\n",
+			    disk->filename, rc, spdk_strerror(rc));
+		disk->unmap = NULL;
+		disk->zero = NULL;
+		return ERROR(rc);
+	}
+
+	if (S_ISBLK(st.st_mode)) {
+		int discard = 0;
+
+		if (ioctl(disk->fd, BLKDISCARDZEROES, &discard) < 0) {
+			rc = errno;
+			SPDK_ERRLOG("ioctl(%s, BLKDISCARDZEROES) failed, errno %d: %s\n",
+				    disk->filename, rc, spdk_strerror(rc));
+			disk->unmap = NULL;
+			disk->zero = NULL;
+			return ERROR(rc);
+		}
+
+		if (discard) {
+			/* rely on fact that device guarantees that discarded blocks will be read back as zeros */
+			SPDK_NOTICELOG("device=%s unmap=discard zero=discard\n", disk->filename);
+			disk->unmap = aio_range_unmap_discard;
+			disk->zero = aio_range_unmap_discard;
+			return 0;
+		}
+
+		/* get kernel to zero blocks on our behalf */
+		SPDK_NOTICELOG("device=%s unmap=none zero=zero\n", disk->filename);
+		disk->unmap = NULL;
+		disk->zero = aio_range_unmap_zero;
 		return 0;
 	}
 
-	rc = close(disk->fd);
-	if (rc < 0) {
-		SPDK_ERRLOG("close() failed (fd=%d), errno %d: %s\n",
-			    disk->fd, errno, spdk_strerror(errno));
+	/* simulate unmap by creating a hole in (sparse) file */
+	disk->unmap = aio_range_fallocate_sparse;
+
+	/* FIXME: this should be configurable */
+
+	if (true) {
+		/* zero blocks by creating a hole in (sparse) file */
+		SPDK_NOTICELOG("file=%s unmap=sparse zero=sparse\n", disk->filename);
+		disk->zero = aio_range_fallocate_sparse;
+		return 0;
+	}
+
+	/* zero blocks by writing zeros */
```

Comment:
Technically this uses the same `blkdev_issue_zeroout()` call that the `BLKZEROOUT` ioctl uses so the kernel tries to issue write zeros to the device first, then falls back to writing zeroes.

Reply:
I think that's fine - that is the behaviour we want.

Reply:
I should have been clearer. The comment is technically correct but it could be clearer by mentioning that how to zero is punted to the kernel (so it decides whether to do it in hardware or software), or something similar.

Reply:
Ok will make the comment a bit clearer.